### PR TITLE
Cms module refactor

### PR DESCRIPTION
**Pull request by SeriousKen**
This is a refactor of the CMS module to move the Twig environment creation into the container avoiding the repetition of creation in the `CmsCompoundObject` and the `Controller`. It removes the controller dependency on the `DebugExtension` which didn't seem necessary and it removes it from the `Cms\Twig\Extension` constructor, which gets it instead from the Twig context.

It shouldn't change any behaviour but this does fix an issue with the Pages plugin where the lack of `DebugExtension` caused [an issue](https://github.com/rainlab/pages-plugin/issues/337)

### CODE REVIEW

File: modules/cms/twig/ComponentNode.php [hunk 1]
```diff
@@ -34,7 +34,7 @@ public function compile(Twig_Compiler $compiler)
         }
 
         $compiler
-            ->write("echo \$this->env->getExtension('Cms\Twig\Extension')->componentFunction(")
+            ->write("echo \$this->env->getExtension('Cms\Twig\Extension')->componentFunction(\$context,")
```
Comment (LukeTowers) [inline modules/cms/twig/ComponentNode.php]: What's the initial backslash on $context for?

Comment (SeriousKen) [reply modules/cms/twig/ComponentNode.php]: Escapes the $ in the double-quoted string, otherwise PHP tries to do variable expansion.

Comment (LukeTowers) [reply modules/cms/twig/ComponentNode.php]: I gotcha. Wouldn't it be simpler to reverse the order of the single and double quotes then? Or am I missing something else?

Comment (SeriousKen) [reply modules/cms/twig/ComponentNode.php]: I just added the $context variable to what was there already. If you swap them wouldn't you then need to double escape the backslashes anyway?

Comment (LukeTowers) [reply modules/cms/twig/ComponentNode.php]: Fair point, that would be the case.

**Comment by LukeTowers** [comment]
@w20k could you review this?

File: modules/cms/ServiceProvider.php [hunk 2]
```diff
@@ -55,6 +63,49 @@ public function boot()
...
+            return new TwigLoader;
+        });
+
+        App::singleton('cms.twig.environment', function ($app) {
+            $loader = $app->make('cms.twig.loader');
+
+            $useCache = !Config::get('cms.twigNoCache');
+            $isDebugMode = Config::get('app.debug', false);
+            $strictVariables = Config::get('cms.enableTwigStrictVariables', false);
+            $strictVariables = is_null($strictVariables) ? $isDebugMode : $strictVariables;
```
Comment (w20k) [inline modules/cms/ServiceProvider.php]: Was there a reason to switch from the short version (`$strictVariables = $strictVariables ?? $isDebugMode;`) to is_null verification?

Comment (LukeTowers) [reply modules/cms/ServiceProvider.php]: @SeriousKen ^

Comment (SeriousKen) [reply modules/cms/ServiceProvider.php]: @w20k to be honest I'm not sure why, but looking at that code `$strictVariables` is never `null`; it's either `true` or `false`, so neither line of code works as I imagine it was intended. Surely it should be `$strictVariables = $strictVariables ?: $isDebugMode`?

Comment (w20k) [reply modules/cms/ServiceProvider.php]: @SeriousKen, could you change that part and also, if you don't mind add few Unit tests?

**Comment by bennothommo** [comment]
Closing - this PR appears to be abandoned.

**Pull request closed without merging**